### PR TITLE
Smtp server detection 1125 v2.0

### PR DESCRIPTION
**Pull request by catenacyber**
Link to ticket: https://redmine.openinfosecfoundation.org/issues/
https://redmine.openinfosecfoundation.org/issues/1125

Describe changes:
- smtp server detection (ie to_client)
- ftp server detection (ie to_client)

SV_BRANCH=https://github.com/OISF/suricata-verify/pull/1850

Follow up on #8892 after merge of https://github.com/OISF/suricata/pull/10982

### CODE REVIEW

**Comment by catenacyber** [comment]
Coming from https://github.com/OISF/suricata/pull/3806 !!!

File: rust/src/util.rs [hunk 1]
```diff
@@ -20,7 +20,103 @@
...
+}
+
+fn parse_domain(input: &[u8]) -> IResult<&[u8], ()> {
+    let (input, _) = many1_count(parse_subdomain)(input)?;
+    let (input, _) = parse_domain_label(input)?;
+    return Ok((input, ()));
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn rs_validate_domain(input: *const u8, in_len: u32) -> u32 {
```
Comment (victorjulien) [inline rust/src/util.rs:59]: I think this supposed to use a different naming style now, right @jasonish?

File: src/app-layer-ftp.c [hunk 2]
```diff
@@ -961,6 +961,35 @@ static AppProto FTPUserProbingParser(
...
+    if (input[3] != ' ' && input[3] != '-') {
+        return ALPROTO_FAILED;
+    }
+    AppProto r = ALPROTO_UNKNOWN;
+    if (f->alproto_ts == ALPROTO_FTP || (f->todstbytecnt > 4 && f->alproto_ts == ALPROTO_UNKNOWN)) {
+        // only validates FTP if client side was FTP
+        // or if client side is unknown despite having received bytes
+        r = ALPROTO_FTP;
+    }
+    for (uint32_t i = 4; i < len; i++) {
```
Comment (victorjulien) [inline src/app-layer-ftp.c:985]: maybe better to use `memchr` for this? Can have optimized implementations.

**Comment by ct0br0** [comment]
qa tokens ran out. ignore the 'pipeline canceled' incoming

File: src/app-layer-ftp.c (hunk 2 above)
Comment (catenacyber) [reply src/app-layer-ftp.c:985]: ok doing

File: rust/src/util.rs (hunk 1 above)
Comment (catenacyber) [reply rust/src/util.rs:59]: Correct, changing

File: src/app-layer-ftp.c (hunk 2 above)
Comment (victorjulien) [reply src/app-layer-ftp.c:985]: probably out of scope for this PR, but there is a rust memchr crate that claims some pretty good performance too IIRC

**Comment by catenacyber** [comment]
Continued in https://github.com/OISF/suricata/pull/11128

**Pull request closed without merging**